### PR TITLE
Fix: language and currency selectors, checkout js bugs and improvements

### PR DESCRIPTION
**Pull request by NeOMakinG**
| Questions         | Answers
| ------------------| -------------------------------------------------------
| Description?      | Language and currency selectors, miss placed div in checkout, checkout back button, first tab rename of checkout personal information, use alert on order confirm message
| Type?             | bug fix
| BC breaks?        | no
| Deprecations?     | no

### CODE REVIEW

**Review by matthieu-rolland — APPROVED** [review]

**Comment by mparvazi** [comment]
The `next-button` should be on the right AND `forward-icon` should be on the right side of `next-button`:

Screenshot 2022-03-29 at 15-19-08 develop

**Comment by mparvazi** [comment]
@NeOMakinG 

I think it's better to show this alert in top of the action buttons:

Screenshot 2022-03-29 at 20-52-55 develop

`templates/checkout/_partials/steps/payment.tpl`

File: src/js/theme.ts [hunk 1]
```diff
@@ -38,14 +38,16 @@ import initResponsiveToggler from './responsive-toggler';
...
+import initCheckout from './pages/checkout';
 import useToast from './components/useToast';
 import useProgressRing from './components/useProgressRing';
 import './modules/blockcart';
 import initProductBehavior from './product';
 import './mobile-menu';
 import './modules/ps_searchbar';
 import './modules/facetedsearch';
+import './modules/ps_languageselector';
+import './modules/ps_currencyselector';
```
Comment (kpodemski) [inline src/js/theme.ts:50]: @NeOMakinG, is there any chance to force webpack to export these files as separated JS files to proper directories? I'm wondering what if one decides to disable these modules. Will we end up having a bunch of not needed JS?

Comment (mparvazi) [reply src/js/theme.ts:50]: @kpodemski maybe someone wants to disable facetedsearch, mobile-menu, ...
I think having just 10 lines of JS code in bundle doesn't impact the performance.

Comment (NeOMakinG) [reply src/js/theme.ts:50]: This is something we need to think about, It should be possible

**Review by matthieu-rolland — APPROVED** [review]

File: templates/checkout/_partials/modal-terms.tpl [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+ * Do not edit or add to this file if you wish to upgrade PrestaShop to newer
+ * versions in the future. If you wish to customize PrestaShop for your
+ * needs please refer to https://devdocs.prestashop.com/ for more information.
+ *
+ * @author    PrestaShop SA and Contributors <contact@prestashop.com>
+ * @copyright Since 2007 PrestaShop SA and Contributors
+ * @license   https://opensource.org/licenses/AFL-3.0 Academic Free License 3.0 (AFL-3.0)
+ *}
+
+ <div id="checkout-modal" class="modal fade" tabindex="-1" role="dialog" aria-labelledby="checkout-modal" aria-hidden="true">
```
Comment (Progi1984) [inline templates/checkout/_partials/modal-terms.tpl:26]:
```suggestion
<div id="checkout-modal" class="modal fade" tabindex="-1" role="dialog" aria-labelledby="checkout-modal" aria-hidden="true">
```

**Review by Progi1984 — COMMENTED** [review]
A small comment

File: src/js/theme.ts (hunk 1 above)
Comment (Hlavtox) [reply src/js/theme.ts:50]: Yes, we need to revert to this. In 1.6, due to separation of css files, only the global.css + bundles related to the controller were loaded. In 1.7, at least modules stayed “modular”.

We cannot make it even worse and make everything into one huge bundle loading everything even if it's not needed. 😬

CCC system is ready for this, it caches unique packs of assets.

**Comment by mparvazi** [comment]
At "final summery":
1. Terms of service is **disabled** in BO
2. No payment method selected
So we will have this alert (but it's wrong):
develop

The alert message should be separated  in context:

develop2

It seems must be fixed here first:
`PS/themes/_core/js/checkout-payment.js`

**Comment by Hlavtox** [comment]
Good find!

@mparvazi What do you think about making the **Place order** button full width, green and with paperplane icon for example?

**Comment by mparvazi** [comment]
@Hlavtox 

We don't need to make it full-width, would be scary in xxlarge screen!
Just floating to the right is enough (in BS standard float-end).

IMO the place order button should has "cart-check" icon:
For example in BS icons:
1. `cart` icon for "Shopping cart"
2. `cart-plus` icon for "Add to cart"
3. `cart-check` icon for "Place order"
Screenshot 2022-04-06 at 14-09-10 Bootstrap Icons

In material-icons we have:
Screenshot 2022-04-06 at 14-14-27 Material Design Icons DX

Btw need to update to latest version for using them ([material-design-icons-iconfont](https://github.com/jossef/material-design-icons-iconfont) v6.5)

**Comment by mparvazi** [comment]
If need to have an icon, doesn't need to has right arrow icon (because it's the last step), 
so my suggestion:
Screenshot 2022-04-06 at 14-26-13 develop

it seems material-icons doesn't have "paper-plane" icon.

**Comment by mparvazi** [comment]
@NeOMakinG 
`
templates/checkout/_partials/order-confirmation-table.tpl`
Ln:116 & Ln:120
`</td>` => `</div>`

File: templates/checkout/_partials/order-confirmation-table.tpl [hunk 3]
```diff
@@ -89,8 +88,7 @@
       </div>
       
       <div class="item__prices col-md-2 col-sm-12 col-12">
-        <div class="text-md-end">{$product.price}</div>
-        <div class="text-md-end">{$product.quantity}</div>
+        <div class="text-md-end">{$product.price} {l s='x%s' sprintf=[$product.quantity] d='Shop.Theme.Catalog'}</div>
```
Comment (mparvazi) [inline templates/checkout/_partials/order-confirmation-table.tpl:91]:
```suggestion
        <div class="text-md-end">({l s='x%s' sprintf=[$product.quantity] d='Shop.Theme.Catalog'}) {$product.price}</div>
```

Comment (mparvazi) [reply templates/checkout/_partials/order-confirmation-table.tpl:91]: Screenshot 2022-04-06 at 14-47-55 develop

File: src/js/modules/ps_currencyselector.ts [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+ *
+ * @author    PrestaShop SA and Contributors <contact@prestashop.com>
+ * @copyright Since 2007 PrestaShop SA and Contributors
+ * @license   https://opensource.org/licenses/AFL-3.0 Academic Free License 3.0 (AFL-3.0)
+ */
+
+const {prestashop} = window;
+
+const initCurrencySelector = () => {
+  const languageSelector = document.querySelector<HTMLElement>('.js-currency-selector');
```
[the remark this answers is not in the document]
Comment (NeOMakinG) [reply src/js/modules/ps_currencyselector.ts:29]: We do, it's a wip

File: src/js/modules/ps_languageselector.ts [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+    const option = event.target as HTMLOptionElement;
+
+    window.location.href = option.value;
+  });
+};
+
+document.addEventListener('DOMContentLoaded', () => {
+  initLanguageSelector();
+
+  prestashop.on('responsive update', () => {
```
[the remark this answers is not in the document]
Comment (NeOMakinG) [reply src/js/modules/ps_languageselector.ts:41]: Seems like you're right, we should rename it

File: src/js/theme.ts (hunk 1 above)
Comment (NeOMakinG) [reply src/js/theme.ts:51]: @jolelievre you got it, we changed the module markup, so we need to handle things with our own markup now

File: src/js/pages/checkout.ts [hunk 6]
```diff
@@ -120,6 +127,42 @@ const initCheckout = () => {
...
+
+              termsModal.show();
+            }
+          } catch (e) {
+            prestashop.emit('handleError', {eventType: 'clickOnTermsLink', error: e});
+          }
+        })();
+      }
+
+      $(prestashop.themeSelectors.modal).modal('show');
```
[the remark this answers is not in the document]
Comment (NeOMakinG) [reply src/js/pages/checkout.ts:163]: Was not needed at all, it was probably a test

**Comment by NeOMakinG** [comment]
> At "final summery":
> 
> 1. Terms of service is **disabled** in BO
> 2. No payment method selected
>    So we will have this alert (but it's wrong):
>    develop
> 
> The alert message should be separated in context:
> 
> develop2
> 
> It seems must be fixed here first: `PS/themes/_core/js/checkout-payment.js`

Do we have a world where `$conditions_to_approve` contains something else than terms? Overwise we can handle it directly in the template

@Hlavtox @mparvazi

**Comment by mparvazi** [comment]
@NeOMakinG 

> Do we have a world where `$conditions_to_approve` contains something else than terms? Overwise we can handle it directly in the template
> 
The `checkout-payment.js` controls the alert container for show/hide, so need to have conditional content for alert:

Screenshot from 2022-04-08 12-40-49

**Comment by NeOMakinG** [comment]
> @NeOMakinG
> 
> > Do we have a world where `$conditions_to_approve` contains something else than terms? Overwise we can handle it directly in the template
> 
> The `checkout-payment.js` controls the alert container for show/hide, so need to have conditional content for alert:
> 
> Screenshot from 2022-04-08 12-40-49

Yeah but as it's an array, I've the feeling that the terms message can be shown even if terms are disabled, with such a condition

**Comment by Hlavtox** [comment]
There is a hook for product terms and conditions. https://github.com/PrestaShop/PrestaShop/issues/26608 Can you try the hook and test the behavior?

**Comment by mparvazi** [comment]
When "Terms of service" in BO is disabled, `$conditions_to_approve|count` returns 0
Btw I've tested this code in several situation and it works.

**Comment by NeOMakinG** [comment]
No one of you responded :p

Is there something else than terms and services in $condition_to_approve? I guess yes as it's an array, then $conditions_to_approve isn't directly pegged to Terms, but can contain another conditions, so we can't use it for this

**Comment by Hlavtox** [comment]
@NeOMakinG `conditions_to_approve` can contain another conditions from hooked modules, so there can be multiple conditions :D

See here - https://github.com/PrestaShop/PrestaShop/blob/develop/classes/checkout/CheckoutPaymentStep.php
And the finder here - https://github.com/PrestaShop/PrestaShop/blob/develop/classes/checkout/ConditionsToApproveFinder.php

**Comment by NeOMakinG** [comment]
Then we need to find another way, like the hook you talked about, I'll search it

**Comment by Hlavtox** [comment]
And if you can make it so it can be optional, it would be good, I would make the core change required. :-)))

So the modules can hook also optional conditions and check their state by JS.

**Comment by NeOMakinG** [comment]
I'm ok with it if we make sure that the theme will be compatible with the latest 1.7 version :p

**Comment by mparvazi** [comment]
> So I think even if the default ones are disabled, there can be conditions from a module.

Is this a problem?
I mean if "Terms of service" in BO is disabled, then other conditions must not be displayed?

**Comment by NeOMakinG** [comment]
> > So I think even if the default ones are disabled, there can be conditions from a module.
> 
> Is this a problem? I mean if "Terms of service" in BO is disabled, then other conditions must not be displayed?

If Terms of service is disabled, it will remove Terms of service from the array, but not others conditions to approve, which aren't forced to be pegged to terms of services, it can be any conditions a module adds

**Comment by mparvazi** [comment]
@NeOMakinG 

To fix undefiend `selectorsMap` in componenets:

Can remove this line: `window.prestashop.themeSelectors = selectorsMap;`
from `theme.ts` and add at the end of `selectors-map.ts`

File: templates/checkout/_partials/modal-terms.tpl [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+       <button type="button" class="btn-close" data-bs-dismiss="modal" aria-label="{l s='Close' d='Shop.Theme.Global'}"></button>
+     </div>
+     <div class="modal-body text-center">
+      <div class="spinner-border text-primary" role="status">
+        <span class="visually-hidden">{l s='Loading...' d='Shop.Theme.Global'}</span>
+      </div> 
+     </div>
+   </div>
+ </div>
+</div>
```
Comment (Progi1984) [inline templates/checkout/_partials/modal-terms.tpl:39]:
```suggestion
<div id="checkout-modal" class="modal fade" tabindex="-1" role="dialog" aria-labelledby="checkout-modal" aria-hidden="true">
  <div class="modal-dialog modal-lg" role="document">
    <div class="modal-content">
      <div class="modal-header">
        <button type="button" class="btn-close" data-bs-dismiss="modal" aria-label="{l s='Close' d='Shop.Theme.Global'}"></button>
      </div>
      <div class="modal-body text-center">
        <div class="spinner-border text-primary" role="status">
          <span class="visually-hidden">{l s='Loading...' d='Shop.Theme.Global'}</span>
        </div> 
      </div>
    </div>
  </div>
</div>
```

**Review by Progi1984 — COMMENTED** [review]
@NeOMakinG Just fixing your indentation in a template.

File: src/js/modules/blockcart.ts [hunk 8]
```diff
@@ -43,23 +46,30 @@ prestashop.blockcart.showModal = (html: string) => {
 
   const mainElement = document.createElement('div');
   mainElement.innerHTML = html;
+  // Select the template of the modal
+  const modalTemplate = document.querySelector<HTMLElement>(BlockcartMap.modal);
```
Comment (mparvazi) [inline src/js/modules/blockcart.ts:50]:
```suggestion
  const modalTemplate = mainElement.querySelector<HTMLElement>(BlockcartMap.modal);
```

File: src/js/theme.ts [hunk 9]
```diff
@@ -37,16 +34,18 @@ import 'bootstrap-input-spinner/src/bootstrap-input-spinner';
 import initResponsiveToggler from './responsive-toggler';
 import initQuickview from './quickview';
 import initCart from './pages/cart';
-import initCheckout from './checkout';
+import initCheckout from './pages/checkout';
 import useToast from './components/useToast';
 import useAlert from './components/useToast';
```
Comment (mparvazi) [inline src/js/theme.ts:39]:
```suggestion
import useAlert from './components/useAlert';
```

Comment (NeOMakinG) [reply src/js/theme.ts:39]: Will be fixed inside another PR

File: templates/checkout/_partials/steps/addresses.tpl [hunk 10]
```diff
@@ -113,15 +113,15 @@
...
 
       <div class="mt-4">
-        <button class="btn btn-outline-primary btn-with-icon d-none d-md-inline-block me-2 js-back" data-step="checkout-personal-information-step">
+        <button class="btn btn-outline-primary btn-with-icon d-block d-md-inline-block me-2 w-full w-md-auto mb-3 mb-md-0 js-back" data-step="checkout-personal-information-step">
           <div class="material-icons">arrow_backward</div>
           {l s='Continue to Personal Information' d='Shop.Theme.Actions'}
         </button>
 
         {if !$form_has_continue_button}
             <button type="submit" class="btn btn-primary btn-with-icon d-block d-md-inline-block w-full w-md-auto continue" name="confirm-addresses" value="1">
```
Comment (mparvazi) [inline templates/checkout/_partials/steps/addresses.tpl:122]:
```suggestion
            <button type="submit" class="btn btn-primary btn-with-icon d-block d-md-inline-block w-full w-md-auto float-end continue" name="confirm-addresses" value="1">
```

File: templates/checkout/_partials/steps/payment.tpl [hunk 11]
```diff
@@ -114,8 +114,23 @@
...
+          '[/2]' => '</a>'
+        ]
+        d='Shop.Theme.Checkout'
+      }
+    </article>
+  {/if}
+
   <div class="payment__actions d-block d-md-flex">
-    <button class="btn btn-outline-primary btn-with-icon d-none d-md-inline-block me-2 js-back" data-step="checkout-delivery-step">
+    <button class="btn btn-outline-primary btn-with-icon d-block d-md-inline-block me-2 w-full w-md-auto mb-3 mb-md-0 js-back" data-step="checkout-delivery-step">
```
Comment (mparvazi) [inline templates/checkout/_partials/steps/payment.tpl:133]:
```suggestion
    <button class="btn btn-outline-primary btn-with-icon d-block d-md-inline-block me-auto w-full w-md-auto mb-3 mb-md-0 js-back" data-step="checkout-delivery-step">
```

File: templates/checkout/_partials/steps/shipping.tpl [hunk 12]
```diff
@@ -100,30 +100,30 @@
...
         </div>
 
         <div class="shipping__actions">
-          <button class="btn btn-outline-primary btn-with-icon d-none d-md-inline-block me-2 js-back" data-step="checkout-addresses-step">
+          <button class="btn btn-outline-primary btn-with-icon d-block d-md-inline-block me-2 w-full w-md-auto mb-3 mb-md-0 js-back" data-step="checkout-addresses-step">
             <div class="material-icons">arrow_backward</div>
             {l s='Back to Addresses' d='Shop.Theme.Actions'}
           </button>
 
           <button type="submit" class="btn btn-primary btn-with-icon d-block d-md-inline-block w-full w-md-auto" name="confirmDeliveryOption" value="1">
```
Comment (mparvazi) [inline templates/checkout/_partials/steps/shipping.tpl:124]:
```suggestion
          <button type="submit" class="btn btn-primary btn-with-icon d-block d-md-inline-block w-full w-md-auto float-end" name="confirmDeliveryOption" value="1">
```

**Review by mparvazi — COMMENTED** [review]
Just a suggestion:
Place the "next buttons" on the right.

Screenshot 2022-04-15 at 23-35-17 develop

File: templates/checkout/_partials/order-final-summary.tpl [hunk 13]
```diff
@@ -22,70 +22,125 @@
...
-          <h4 class="h5 black addresshead">{l s='Your Delivery Address' d='Shop.Theme.Checkout'}</h4>
-          {$customer.addresses[$cart.id_address_delivery]['formatted'] nofilter}
+<hr />
+<section id="order-summary-content" class="order__summary">
+  <h4 class="mb-4">{l s='Please check your order before payment' d='Shop.Theme.Checkout'}</h4>
+
+  <div class="order__summary__addresses mb-4">
+    <h5 class="mb-3">
+      {l s='Addresses' d='Shop.Theme.Checkout'}
+      <span class="step-edit step-to-addresses fs-6 text-gray js-edit-addresses" data-step="checkout-addresses-step"><i class="material-icons edit">mode_edit</i> {l s='edit' d='Shop.Theme.Actions'}</span>
```
Comment (mparvazi) [inline templates/checkout/_partials/order-final-summary.tpl:33]:
```suggestion
      <span class="btn step-edit step-to-addresses fs-6 text-gray js-edit-addresses" data-step="checkout-addresses-step"><i class="material-icons fs-6 edit">mode_edit</i> {l s='Edit' d='Shop.Theme.Actions'}</span>
```

File: templates/checkout/_partials/order-final-summary.tpl [hunk 14]
```diff
@@ -22,70 +22,125 @@
...
     </div>
   </div>
 
   {if !$cart.is_virtual}
-    <h4 class="h4">
+    <h5 class="mb-3">
       {l s='Shipping Method' d='Shop.Theme.Checkout'}
-      <span class="step-edit step-to-delivery js-edit-delivery"><i class="material-icons edit">mode_edit</i> {l s='edit' d='Shop.Theme.Actions'}</span>
-    </h4>
+      <span class="step-edit step-to-addresses fs-6 text-gray js-edit-shipping" data-step="checkout-delivery-step"><i class="material-icons edit">mode_edit</i> {l s='edit' d='Shop.Theme.Actions'}</span>
```
Comment (mparvazi) [inline templates/checkout/_partials/order-final-summary.tpl:98]:
```suggestion
      <span class="btn step-edit step-to-addresses fs-6 text-gray js-edit-shipping" data-step="checkout-delivery-step"><i class="material-icons fs-6 edit">mode_edit</i> {l s='Edit' d='Shop.Theme.Actions'}</span>
```

**Review by mparvazi — COMMENTED** [review]
Fix "Edit button" for sections on "order final summary" page.

Screenshot 2022-04-15 at 23-51-49 develop

File: templates/checkout/_partials/order-final-summary.tpl [hunk 15]
```diff
@@ -22,70 +22,125 @@
...
+      <div class="col-12 col-sm-6 mb-2">
+        <div class="address card">
+          <div class="card-body">
+            <div class="address__content col-10">
+              <h6 class="address__alias card-title">{l s='Your Delivery Address' d='Shop.Theme.Checkout'}</h6>
+              <address class="address__content">{$customer.addresses[$cart.id_address_delivery]['formatted'] nofilter}</address>
+
+              <div class="address__actions">
+                <a
+                  class="address__edit text-muted ps-0 pb-0"
```
Comment (mparvazi) [inline templates/checkout/_partials/order-final-summary.tpl:46]:
```suggestion
                  class="address__edit ps-0 pb-0"
```

File: templates/checkout/_partials/order-final-summary.tpl [hunk 16]
```diff
@@ -22,70 +22,125 @@
...
+      <div class="col-12 col-sm-6 mb-2">
+        <div class="address card">
+          <div class="card-body">
+            <div class="address__content col-10">
+              <h6 class="address__alias card-title">{l s='Your Invoice Address' d='Shop.Theme.Checkout'}</h6>
+              <address class="address__content">{$customer.addresses[$cart.id_address_delivery]['formatted'] nofilter}</address>
+
+              <div class="address__actions">
+                <a
+                  class="address__edit text-muted ps-0 pb-0"
```
Comment (mparvazi) [inline templates/checkout/_partials/order-final-summary.tpl:74]:
```suggestion
                  class="address__edit ps-0 pb-0"
```

**Review by mparvazi — COMMENTED** [review]
We should not use "text-muted" for buttons or links. The hover state will be off for link (text-muted uses !important for color in BS).
```
.text-muted {
  color: #6c757d !important;
}
```

File: templates/checkout/_partials/steps/addresses.tpl (hunk 10 above)
Comment (Hlavtox) [reply templates/checkout/_partials/steps/addresses.tpl:122]: I would use flex, justify content space between. The floats sometime cause weird parent div height behavior. 👍

**Comment by NeOMakinG** [comment]
Fixed last feedback, can we finally merge this? So we can provide every fixes to the theme :p

**Comment by NeOMakinG** [comment]
As noone from @mparvazi and @Hlavtox disapproved, I'll merge this one and iterate on a new PR in order to fix last fixes from the big issue list

**Pull request merged**